### PR TITLE
fix(convo_miner): enforce CHUNK_SIZE in paragraph chunker (#1534)

### LINKED ISSUE #1534: convo_miner._chunk_by_paragraph does not enforce CHUNK_SIZE, causing oversized chunks that crash embedding on conversations containing large pasted content
## Summary

`mempalace mine <dir> --mode convos` can crash with `RuntimeError: Invalid buffer size: 120.00 GiB in upsert.` when a session transcript contains a single very-long line (e.g., a large pasted document, a multi-hundred-KB tool output) that gets routed through `_chunk_by_paragraph` (the fallback chunker) instead of `_chunk_by_exchange`. The fallback chunker doesn't enforce `CHUNK_SIZE`, so the oversized paragraph becomes a single oversized chunk that exceeds the embedding model's attention memory budget on `collection.upsert(...)`.

`_chunk_by_exchange` handles size correctly (splits content above `CHUNK_SIZE` across drawers). `_chunk_by_paragraph` and its line-group fallback do not. The fix is mechanical: mirror the splitting loop into both paragraph paths.

## Reproduction

**Trigger I hit:**

```bash
mempalace mine ~/.claude/projects/ --mode convos --wing claude-sessions
```

Where `~/.claude/projects/` contained ~247 Claude Code session `.jsonl` files. Two recent files contained pasted content as single very-long lines:
- File A: 2.88 MB total, max line length 135,600 chars
- File B: 2.10 MB total, max line length 119,671 chars

**Observed:**

```
File "mempalace/convo_miner.py", line 489, in mine_convos
    drawers_added, room_delta, skipped = _file_chunks_locked(...)
File "mempalace/convo_miner.py", line 370, in _file_chunks_locked
    collection.upsert(documents=batch_docs, ids=batch_ids, metadatas=batch_metas)
…
File ".../transformers/integrations/sdpa_attention.py", line 92, in sdpa_attention_forward
    attn_output = torch.nn.functional.scaled_dot_product_attention(...)
RuntimeError: Invalid buffer size: 120.00 GiB in upsert.
```

The 120 GiB number is O(seq_len²) attention-matrix memory — one chunk in the failing batch has a sequence length large enough to blow the matrix.

**Smoking-gun comparison via `--dry-run`:**

Three same-source-format `.jsonl` files of comparable wall-clock size:

| File | Path through chunker | Drawer count from dry-run | Implied avg chunk size |
|---|---|---|---|
| Older session, normalizes to ≥3 `>` lines | `_chunk_by_exchange` (CHUNK_SIZE-bounded) | **1352** | ~2 KB |
| File A above | `_chunk_by_paragraph` (NOT CHUNK_SIZE-bounded) | **170** | ~17 KB |
| File B above | `_chunk_by_paragraph` (NOT CHUNK_SIZE-bounded) | **135** | ~15 KB |

The ratio is the bug — same path, same source format, but the dispatch in `chunk_exchanges` (line 108) sends some files down the unsafe fallback when `quote_lines >= 3` doesn't hold for the normalized content.

## Root cause

`mempalace/convo_miner.py`:

**Path 1 — `_chunk_by_paragraph` paragraph loop (lines 184-186):**

```python
for para in paragraphs:
    if len(para) > MIN_CHUNK_SIZE:
        chunks.append({"content": para, "chunk_index": len(chunks)})  # ← no CHUNK_SIZE cap
```

**Path 2 — `_chunk_by_paragraph` line-group fallback (lines 178-181):**

```python
if len(paragraphs) <= 1 and content.count("\n") > 20:
    lines = content.split("\n")
    for i in range(0, len(lines), 25):
        group = "\n".join(lines[i : i + 25]).strip()
        if len(group) > MIN_CHUNK_SIZE:
            chunks.append({"content": group, "chunk_index": len(chunks)})  # ← no CHUNK_SIZE cap
```

**For contrast — `_chunk_by_exchange` does it correctly (lines 145-156):**

```python
if len(content) > CHUNK_SIZE:
    first_part = content[:CHUNK_SIZE]
    if len(first_part.strip()) > MIN_CHUNK_SIZE:
        chunks.append({"content": first_part, "chunk_index": len(chunks)})
    remainder = content[CHUNK_SIZE:]
    while remainder:
        part = remainder[:CHUNK_SIZE]
        remainder = remainder[CHUNK_SIZE:]
        if len(part.strip()) > MIN_CHUNK_SIZE:
            chunks.append({"content": part, "chunk_index": len(chunks)})
```

## Suggested fix (sketch)

A small helper to mirror the bounded-splitting pattern from `_chunk_by_exchange`:

```python
def _emit_bounded(chunks: list, content: str) -> None:
    """Append content as one or more chunks, none exceeding CHUNK_SIZE."""
    while content:
        part = content[:CHUNK_SIZE]
        content = content[CHUNK_SIZE:]
        if len(part.strip()) > MIN_CHUNK_SIZE:
            chunks.append({"content": part, "chunk_index": len(chunks)})
```

Then replace both unsafe sites with `_emit_bounded(chunks, group)` and `_emit_bounded(chunks, para)` respectively.

Maintainers may prefer to inline the loop or rename — the shape matters more than the exact factoring.

## Environment

- mempalace: HEAD at `604df3b` (one local-only commit ahead of `origin/main` raising `MAX_CHUNKS_PER_FILE` 500→50_000, which lives in `miner.py` not `convo_miner.py` and is unrelated to this bug)
- Python 3.13, macOS Apple Silicon
- Embedding model: xlm-roberta variant (391 weights, default mempalace stack)
- Palace: ChromaDB backend, default config

## Scope

- Affects: `mempalace mine --mode convos` against any directory containing files whose `normalize()` output lacks ≥3 `>`-prefixed lines AND contains paragraphs larger than `CHUNK_SIZE`. Claude Code session transcripts with large pasted content / large tool outputs hit this regularly.
- Does NOT affect: `--mode projects` (different chunker in `miner.py`; haven't audited but the dispatch is separate), files that route through `_chunk_by_exchange`, or storage / search / hooks layers.
- Severity: medium-high. Failure is loud (exception, not silent corruption) but stops a long-running operation cold and consumes the embedding-model setup cost before crashing.

## What this is NOT

- NOT a storage corruption bug — previously-stored drawers are intact.
- NOT a re-mining duplicates bug — `file_already_mined` (line 429) skips already-ingested files; duplicate "already exists" errors at upsert (lines 376-378) are caught.
- NOT related to `MAX_CHUNKS_PER_FILE` (`miner.py` constant; not in `convo_miner.py`'s call graph).

### PR DESCRIPTION
**Pull request by mvalentsev**
## What does this PR do?

Fixes #1534. `mempalace mine --mode convos` crashed with `RuntimeError: Invalid buffer size: 120.00 GiB in upsert.` whenever a transcript routed through `_chunk_by_paragraph` (the fallback chunker, invoked when fewer than 3 `>`-prefixed lines are present in the normalized content) contained a paragraph or 25-line group longer than `CHUNK_SIZE`. Both unsafe sites in `_chunk_by_paragraph` appended whole paragraphs or line groups as single drawers with no per-drawer size cap. ChromaDB's embedding model then attempted an O(seq_len²) attention allocation on the oversized chunk and crashed.

For Claude Code users mining their session history (one of the primary use cases for convos mode), this blocks ingestion entirely once any oversized-chunk file enters the corpus.

The fix extracts a `_emit_bounded` helper that mirrors the continuation-slice loop already used in `_chunk_by_exchange`, and calls it from both unsafe sites in `_chunk_by_paragraph`. The signature of `_chunk_by_paragraph` gains a `chunk_size` parameter so the config-resolved value from `chunk_exchanges` reaches the paragraph path (previously only `min_chunk_size` was plumbed, leaving `MempalaceConfig.chunk_size` ineffective on this path post-#1024). The two magic numbers in the line-group fallback (`25` group size, `20` newline threshold) are extracted as `_LINE_GROUP_SIZE` and `_LINE_FALLBACK_MIN_NEWLINES` while in the area.

Reporter @davidglidden provided a high-quality bug report including the exact failing path, a fix sketch with the `_emit_bounded` helper name and shape, and a follow-up corpus analysis across 247 files. The implementation adopts that suggestion; credit goes via Co-Authored-By.

**Coordination note** for in-flight PRs touching the same area:
- #1449 (verbatim whitespace preservation) modifies the same two sites this PR replaces. Both intents are compatible: after this PR, the helper takes raw paragraph/group content and slices verbatim; #1449's intent of preserving leading whitespace can be folded by dropping the `.strip()` in the paragraph comprehension. Happy to coordinate rebase order with whichever lands first.
- #1137 (verbatim newlines in `_chunk_by_exchange`) touches a different function, no overlap.

**Audit findings (not in this PR's scope; separate follow-up issues will be filed):**
- `mempalace/general_extractor.py:413-419` (`extract_memories`): same class of bug. Appends `_split_into_segments` outputs as memories without a size cap. Different semantic surface (each segment has its own scoring and disambiguation, so post-hoc slicing would change classification), so the fix shape differs from this PR.
- `mempalace/diary_ingest.py:164` (`drawers_col.upsert(documents=[text]...)`): same class of bug. A multi-MB daily diary file would crash embedding upsert the same way. Different entrypoint (`mempalace sync` daily diary), likely wants entry-aware chunking rather than character slicing.

**Backwards compat note:** previously-mined files that were broken on `develop` will not auto-rechunk after this PR (they remain marked `file_already_mined`). Users who want to retroactively chunk old oversized drawers can `mempalace repair` or re-mine with `--force` (out of scope here).

## How to test

Manual reproducer for the original crash on `develop` tip `cabc3af`:

```bash
TMPDIR=$(mktemp -d)
uv run python -c "open('$TMPDIR/giant.txt','w').write('x' * 200000)"
mempalace mine "$TMPDIR" --mode convos --wing repro-1534
# develop:    RuntimeError: Invalid buffer size: ... GiB in upsert
# this branch: exit 0, 250 drawers filed (200000 / CHUNK_SIZE=800)
```

End-to-end smoke test on four pathological shapes (paragraph-loop + line-group fallback + huge-single-line + clean-exchange):

```bash
TMPDIR=$(mktemp -d)
uv run python -c "
open('$TMPDIR/para_big.txt','w').write('A' * 5000 + chr(10)*2 + 'small paragraph')
open('$TMPDIR/lines_big.txt','w').write(chr(10).join(['x' * 40] * 60))
open('$TMPDIR/huge_line.txt','w').write('Z' * 134000)
open('$TMPDIR/exchange.txt','w').write('> q\nresponse\n\n> q\nresponse\n\n> q\nresponse')
"
mempalace mine "$TMPDIR" --mode convos --wing repro-1534-mixed
# Expected on this branch: 183 drawers filed across 4 files, exit 0,
# no RuntimeError from the embedding path.
```

Unit suite (TDD red on develop, green here):

```bash
uv run pytest tests/test_convo_miner_unit.py -v
```

Coverage added (13 new + 1 updated test across `TestChunkExchanges` and new `TestEmitBounded`):
- `_chunk_by_paragraph` paragraph-loop and line-group-fallback both enforce `CHUNK_SIZE`
- custom `chunk_size` via `chunk_exchanges(..., chunk_size=N)` propagates to the paragraph path
- verbatim invariant: joined chunks equal source (no content loss, in source order)
- explicit slice-order test (`"a"*CHUNK_SIZE + "b"*CHUNK_SIZE + "c"*CHUNK_SIZE` chunks in that order)
- sub-`MIN_CHUNK_SIZE` trailing groups dropped
- oversized boundary, many-multiples, and exact-CHUNK_SIZE edge cases pinned
- direct unit tests for `_emit_bounded` (empty content, whitespace-only slice, chunk-index continuity, gate enforcement)

## Checklist
- [x] Tests pass (`uv run pytest tests/ -v --ignore=tests/benchmarks`; 1 pre-existing unrelated flake in `test_palace_locks::test_same_palace_serializes_across_processes` passes in isolation)
- [x] No hardcoded paths
- [x] Linter passes (`uvx --from 'ruff>=0.4.0,<0.5' ruff check . && uvx --from 'ruff>=0.4.0,<0.5' ruff format --check .` per CI ruff version pin)

### CODE REVIEW

File: mempalace/convo_miner.py [hunk 1]
```diff
@@ -190,23 +192,42 @@ def _chunk_by_exchange(lines: list, chunk_size: int, min_chunk_size: int) -> lis
...
+    paragraph fallback honors the same per-drawer size cap. Each emitted
+    slice is gated by ``len(part.strip()) > min_chunk_size``, so slices
+    whose stripped length does not exceed the floor (including all-whitespace
+    tails left by mid-content splits) are dropped.
+    """
+    while content:
+        part = content[:chunk_size]
+        content = content[chunk_size:]
+        if len(part.strip()) > min_chunk_size:
+            chunks.append({"content": part, "chunk_index": len(chunks)})
```
[the remark this answers is not in the document]
Comment (mvalentsev) [reply mempalace/convo_miner.py]: Thanks. Applied the index-based loop in 66b5194: `range(0, len(content), chunk_size)` with `content[i : i + chunk_size]`. Drops the O(N²) substring-copy pattern, and lets `_chunk_by_exchange` reuse the same primitive (both call sites now delegate to `_emit_bounded`).

File: mempalace/convo_miner.py [hunk 2]
```diff
@@ -190,23 +192,42 @@ def _chunk_by_exchange(lines: list, chunk_size: int, min_chunk_size: int) -> lis
...
+    Mirrors the continuation-slice loop used in ``_chunk_by_exchange`` so the
+    paragraph fallback honors the same per-drawer size cap. Each emitted
+    slice is gated by ``len(part.strip()) > min_chunk_size``, so slices
+    whose stripped length does not exceed the floor (including all-whitespace
+    tails left by mid-content splits) are dropped.
+    """
+    while content:
+        part = content[:chunk_size]
+        content = content[chunk_size:]
+        if len(part.strip()) > min_chunk_size:
```
[the remark this answers is not in the document]
Comment (mvalentsev) [reply mempalace/convo_miner.py]: Confirmed and fixed in 66b5194. The `min_chunk_size` floor now gates the WHOLE stripped content (noise filter on the input), and every slice is emitted verbatim. New test `test_split_805_chars_at_chunk_size_800_preserves_tail` directly pins the case you raised: 805 chars + chunk_size=800 + min_chunk_size=30 → 2 drawers (800 + 5), tail preserved. The same gate change applies to `_chunk_by_exchange` which had the identical per-slice gate.

File: tests/test_convo_miner_unit.py [hunk 3]
```diff
@@ -97,6 +120,109 @@ def test_long_ai_response_not_truncated(self):
...
+        assert max_len <= 400, f"oversized chunk under custom chunk_size=400: {max_len}"
+
+    def test_paragraph_loop_no_content_loss(self):
+        """Verbatim principle: every char of a single long paragraph lands
+        in some drawer in order. The slicing helper must not drop or
+        reorder content."""
+        content = "a" * 5000
+        chunks = chunk_exchanges(content)
+        joined = "".join(c["content"] for c in chunks)
+        assert joined == content
```
[the remark this answers is not in the document]
Comment (mvalentsev) [reply tests/test_convo_miner_unit.py:160]: Good catch. The whole-content gate in 66b5194 makes the verbatim assertion robust regardless of tail length (5000 mod 800 = 200 is no longer load-bearing). Added an explicit `test_split_805_chars_at_chunk_size_800_preserves_tail` for the edge case you cited, plus `test_whole_content_below_floor_dropped` to pin the noise-filter contract.
